test/e2e/network/service.go

```diff
@@ -518,6 +518,123 @@ func pokeUDP(host string, port int, request string, params *UDPPokeParams) UDPPo
 	return ret
 }
 
+// continuousEcho() uses the same connection for multiple requests, made to run as a goroutine so that
+// manipulations can be made to the service and backend pods while a connection is ongoing
+// it starts by sending a series of packets to establish conntrack entries and waits for a signal to keep
+// sending packts. It returns an error if the number of failed attempts is >= 5
+func continuousEcho(host string, port int, timeout time.Duration, maxAttempts int, signal chan struct{}, errorChannel chan error) {
+	defer ginkgo.GinkgoRecover()
+	const threshold = 10
+
+	// Sanity check inputs, because it has happened.  These are the only things
+	// that should hard fail the test - they are basically ASSERT()s.
+	if host == "" {
+		errorChannel <- fmt.Errorf("Got empty host for continuous echo (%s)", host)
+		return
+	}
+	if port == 0 {
+		errorChannel <- fmt.Errorf("Got port ==0 for continuous echo (%d)", port)
+		return
+	}
+
+	hostPort := net.JoinHostPort(host, strconv.Itoa(port))
+	url := fmt.Sprintf("udp://%s", hostPort)
+
+	ret := UDPPokeResult{}
+
+	con, err := net.Dial("udp", hostPort)
+	if err != nil {
+		ret.Status = UDPError
+		ret.Error = err
+		errorChannel <- fmt.Errorf("Connection to %q failed: %v", url, err)
+		return
+	}
+
+	numErrors := 0
+
+	for i := 0; i < maxAttempts; i++ {
+		if i == threshold {
+			framework.Logf("Continuous echo waiting for signal to continue")
+			<-signal
+			if numErrors == threshold {
+				errorChannel <- fmt.Errorf("continuous echo was not able to communicate with initial server pod")
+				return
+			}
+		}
+		time.Sleep(1 * time.Second)
+		err = con.SetDeadline(time.Now().Add(timeout))
+		if err != nil {
+			ret.Status = UDPError
+			ret.Error = err
+			framework.Logf("Continuous echo (%q): %v", url, err)
+			numErrors++
+			continue
+		}
+		myRequest := fmt.Sprintf("echo %d", i)
+		_, err = con.Write([]byte(fmt.Sprintf("%s\n", myRequest)))
+		if err != nil {
+			ret.Error = err
+			neterr, ok := err.(net.Error)
+			if ok && neterr.Timeout() {
+				ret.Status = UDPTimeout
+			} else if strings.Contains(err.Error(), "connection refused") {
+				ret.Status = UDPRefused
+			} else {
+				ret.Status = UDPError
+			}
+			numErrors++
+			framework.Logf("Continuous echo (%q): %v - %d errors seen so far", url, err, numErrors)
+			continue
+		}
+
+		err = con.SetDeadline(time.Now().Add(timeout))
+		if err != nil {
+			ret.Status = UDPError
+			ret.Error = err
+			numErrors++
+			framework.Logf("Continuous echo (%q): %v - %d errors seen so far", url, err, numErrors)
+			continue
+		}
+
+		bufsize := len(strconv.Itoa(i)) + 1
+		var buf = make([]byte, bufsize)
```

Comment:
just curious, why do we have to recreate the buffer and increase the size in each iteration?

Reply:
oversight from when I created this function. Moved it to outside the for loop